### DISCUSSION TITLE
Inconsistent setting of errors in FormState.useSubmit() callback

### DISCUSSION THREAD

**Discussion by kylebjordahl** [ariakit/ariakit#2362] [Questions]
I'm running into some inconsistent behavior when setting errors in either `useValidate` or `useSubmit` callbacks; specifically, when setting errors via a submit handler, I'm seeing cases where the form appears to clear it's own values despite me calling `setErrors` in the submit handler. The only way I can consistently avoid this is by throwing an error in the handler, which feels wrong (and also results in a console error which is unacceptable for our use case).

Could we possibly implement a return value from the useSubmit handler that would indicate a failure to submit? It looks like the submit method is trying to detect any errors which were set during the callback runtime, but its unclear to me why it's failing to catch my errors. If I log the form state in a useEffect for every time the form state changes, I can see a couple updates where the error is set properly, but then one or two of the last updates show no errors and no form field values.

This also got me wondering if there is a particular reason that the submit and validate callbacks don't provide their own reference to the current state of the form? I'm constantly finding myself wanting to wrap my own callback functions in `useCallback` but I'm not sure that will work correctly, given the context in which the callback runs.

If there's not a strong design reason for this, would the maintainers be open to a PR to add something like this?

```ts
useValidate: (callback:(formState: FormState<V>)) => void
```

**Comment by diegohaz** [comment]
Hi @kylebjordahl!

Could you please create an issue with a reproduction? I'm not even sure which version you're talking about. So it's hard to help.

**Reply by kylebjordahl** [reply]
Totally! Wasn't sure if this warranted an issue since it might just be me doing something wrong. Also, thanks for the super fast reply @diegohaz !

**Reply by kylebjordahl** [reply]
So, I was unable to reproduce minimally, which points to some other issue in my specific use I suppose; I did however discover the `resetOnSubmit` prop, which works great for me.

I also noticed that in the form example in the docs, the form state does appear to be sent to the callbacks, I just don't see it in the types for `^2.0.0-next.41` so I'm assuming that is a recent addition. With any luck, I'll have time to offer a PR for the docs to cover that for someone like me in the future.
